syntax = "proto3";
package ch_ewf.psc;

import "google/protobuf/timestamp.proto";
import "base_types.proto";

message Notification {
  oneof psc {
    CorporateEntity corporate = 1;
    LegalPerson legal_person = 2;
    Individual individual = 3;
  }
}

message CorporateEntity {
  string corporate_name = 1;
  base_types.CompanyAddress address = 2;
  CorporateIdentification corporate_identification = 3;
}

message CorporateIdentification {
  string place_registered = 1;
  string registration_number = 2;
  string law_governed = 3;
  string legal_form = 4;
  string country_or_state = 5;
}

message Individual {
  base_types.PersonName person = 1;
  base_types.ServiceAddress service_address = 2;
  google.protobuf.Timestamp date_of_birth = 3;
  string nationality = 4;
  string country_of_residence = 5;
  base_types.ResidentialAddress residential_address = 6;
  bool consent_statement = 7;
}

message IndividualIdentification {
  base_types.PersonName name = 1;
  base_types.PartialDOB partial_dob = 2;
}

message LegalPerson {
  string name = 1;
  base_types.CompanyAddress address = 2;
  LegalPersonIdentification legal_person_identification = 3;
}

message LegalPersonIdentification {
  string law_governed = 1;
  string legal_form = 2;
}

message NatureOfControls {
  oneof nature_of_controls {
    CompanyNatureOfControls company_nature_of_controls = 4;
    LLPNatureOfControls llp_nature_of_controls = 5;
  }
}


message CompanyNatureOfControls {
  enum NatureOfControl {
    OwnershipOfShares25To50 = 0;
    OwnershipOfShares50To75 = 1;
    OwnershipOfShares75To100 = 2;
    OwnershipOfShares25To50AsTrust = 3;
    OwnershipOfShares50To75AsTrust = 4;
    OwnershipOfShares75To100AsTrust = 5;
    OwnershipOfShares25To50AsFirm = 6;
    OwnershipOfShares50To75AsFirm = 7;
    OwnershipOfShares75To100AsFirm = 8;
    VotingRights25To50 = 9;
    VotingRights50To75 = 10;
    VotingRights75To100 = 11;
    VotingRights25To50AsTrust = 12;
    VotingRights50To75AsTrust = 13;
    VotingRights75To100AsTrust = 14;
    VotingRights25To50AsFirm = 15;
    VotingRights50To75AsFirm = 16;
    VotingRights75To100AsFirm = 17;
    RightToAppointAndRemoveDirectors = 18;
    RightToAppointAndRemoveDirectorsAsTrust = 19;
    RightToAppointAndRemoveDirectorsAsFirm = 20;
    SignificantInfluence = 21;
    SignificantInfluenceAsTrust = 22;
    SignificantInfluenceAsFirm = 23;
  }

  repeated NatureOfControl nature_of_controls = 1;
}

message LLPNatureOfControls {
  enum NatureOfControl {
    RightToSurplusAssets25To50 = 0;
    RightToSurplusAssets50To75 = 1;
    RightToSurplusAssets75To100 = 2;
    RightToSurplusAssets25To50AsTrust = 3;
    RightToSurplusAssets50To75AsTrust = 4;
    RightToSurplusAssets75To100AsTrust = 5;
    RightToSurplusAssets25To50AsFirm = 6;
    RightToSurplusAssets50To75AsFirm = 7;
    RightToSurplusAssets75To100AsFirm = 8;
    VotingRights25To50 = 9;
    VotingRights50To75 = 10;
    VotingRights75To100 = 11;
    VotingRights25To50AsTrust = 12;
    VotingRights50To75AsTrust = 13;
    VotingRights75To100AsTrust = 14;
    VotingRights25To50AsFirm = 15;
    VotingRights50To75AsFirm = 16;
    VotingRights75To100AsFirm = 17;
    RightToAppointAndRemoveMembers = 18;
    RightToAppointAndRemoveMembersAsTrust = 19;
    RightToAppointAndRemoveMembersAsFirm = 20;
    SignificantInfluence = 21;
    SignificantInfluenceAsTrust = 22;
    SignificantInfluenceAsFirm = 23;
  }

  repeated NatureOfControl nature_of_controls = 1;
}


  enum PSCLevelStatement {
    ExistsButNotIdentified = 0;
    DetailsNotConfirmed = 1;
    ContactedButNoResponse = 2;
    RestrictionNoticeIssued = 3;
  }

  enum CompanyLevelStatement {
  NoSignificantControl = 0;
  StepsNotCompleted = 1;
}

message StatementNotification {
  oneof notification {
    CompanyLevelStatement company_level_statement = 1;
    PSCLevelStatement psc_level_statement = 2;
    LinkedStatement linked_statement = 3;
  }
}

message LinkedStatement {
  enum LinkedStatementType {
    FailedToConfirmChangedDetails = 0;
  }

  LinkedStatementType statement = 1;
  oneof psc {
    IndividualIdentification individual = 2;
    string corporate_name = 3;
    string legal_person_name = 4;
    bool super_secure_individual = 5;
  }
}